pkg/runtime/runtime.go

```diff
@@ -571,7 +572,13 @@ func (a *DaprRuntime) initPluggableComponents() {
 
 // Sets the status of the app to healthy or un-healthy
 // Callback for apphealth when the detected status changed
-func (a *DaprRuntime) appHealthChanged(status uint8) {
+func (a *DaprRuntime) appHealthChanged(status apphealth.AppHealthStatus) {
+	if a.appHealthStatus == status {
```

Comment:
Can you use an atomic flag here for a.appHealthStatus? Otherwise this could cause race conditions

Reply:
https://github.com/dapr/dapr/blob/master/pkg/apphealth/health.go#L195

It will not appear here, the current detection interval is the shortest 1s, and it is executed serially, so the callback API  changes the status, and there will be cause race conditions.

Reply:
Although the interval is 1s, they are invoked as background goroutines (https://github.com/dapr/dapr/blob/master/pkg/apphealth/health.go#L205) so depending on how busy the application is, they may be delayed in their execution

Reply:
Does it mean that the goroutine is heavily blocked? If the goroutine blocks for more than the interval period, there will indeed be a race condition.

But there is another point here: if the later request arrives first, the first request arrives later, and then the Callback  is executed, is the state itself wrong? Shouldn't that goroutine be removed? become blocking execution?

Reply:
> But there is another point here: if the later request (goroutine) arrives first, the first request (goroutine) arrives later, and then the Callback is executed, is the state itself wrong (maybe)? Shouldn't that goroutine be removed? become blocking execution?

That's a good point.

Actually, I do first wonder why is the callback invoked twice. Rather than adding a check to the runtime, shouldn't there be a check in the apphealth package to make sure that the change is never sent to the runtime if it's not actually changed in the first place?